### ISSUE TITLE
iou loss, pred and weight ndim not equal issue

### ISSUE THREAD

**Issue by ceykmc** [open-mmlab/mmdetection#4456]
in mmdet/models/losses/iou_loss.py, exist 5 lines of:
`return (pred * weight).sum()  # 0`
but if pred ndim is not equal to weight ndim, they can not multiply.
it happened when i train gfl detector.

with the suggestion of @shinya7y  
should replace
```
return (pred * weight).sum()  # 0
```
with
```
if pred.dim() == weight.dim() + 1:  
    weight = weight.unsqueeze(1)
return (pred * weight).sum()  # 0
```
??

**Comment by ZwwWayne** [comment]
Hi @ceykmc ,
Could you provide a way for us to reproduce your bug?

**Comment by ceykmc** [comment]
in gfl_head.py,  ```loss_single``` function

```weight_targets = cls_score.detach().sigmoid()```
```
loss_bbox = self.loss_bbox(
                pos_decode_bbox_pred,
                pos_decode_bbox_targets,
                weight=weight_targets,
                avg_factor=1.0)
```
in the corner case: all elements in weight_targets is zero.
This may happen in the early training period.

then, in the iou_loss.py, GIoULoss forward function,
```
if weight is not None and not torch.any(weight > 0):
            return (pred * weight).sum()  # 0
```
the pred is 2 dimention, weight is 1 dimention, they can not multiply

**Comment by shinya7y** [comment]
Unstable training will cause this RuntimeError (possibly as a premonitory phenomenon of nan).

COCO from scratch with high learning rates (according to my old memo)
- SyncBN, lr0.04
- GN+WS, lr0.02

Custom datasets (some settings may be inappropriate)
https://github.com/implus/GFocal/issues/10
https://github.com/shinya7y/UniverseNet/issues/7

**Comment by dave-andersen** [comment]
Confirming the problem - I'm encountering it regularly. The proposed fix in the first comment doesn't work, though: While it corrects the shape mismatch in the multiplication, the training ends up full of NaN's and things fail from there from a training perspective, as @shinya7y notes.

I'm encountering it in a dataset with fairly sparse occurrences of the target class, with large images and cropping, such that cropping may produce images with no detection targets whatsoever. I suspect but haven't yet confirmed this may be causing the problem for me.

**Comment by dave-andersen** [comment]
For others, I've temporarily hacked around training going bad by skipping the bad examples:
gfl_head.py line 250:
if len(pos_inds) > 0 and all(cs > -3e20 for cs in cls_score):

This is a bad solution and I'm probably a bad person for hacking it this way instead of digging for the root cause more, but hopefully this is at least a starting point and lets folks get a little work done in the meantime.

**Comment by dave-andersen** [comment]
For others, I've temporarily hacked around training going bad by skipping the bad examples:
gfl_head.py line 250:
if len(pos_inds) > 0 and all(cs > -3e20 for cs in cls_score):

This is a bad solution and I'm probably a bad person for hacking it this way instead of digging for the root cause more, but hopefully this is at least a starting point and lets folks get a little work done in the meantime. I suspect the real underlying cause is when a model is starting to not train and going crazy, thus producing very bad scores, but this one has been tricky to debug. *I do not recommend the above hack as a long-term solution* but give it a try and if it works it'll be useful to know for debugging further.

**Comment by shinya7y** [comment]
Please refer to https://github.com/open-mmlab/mmdetection/pull/4978

**Comment by hhaAndroid** [comment]
In any case, the problem of `(pred * weight).sum()` error still needs to be fixed, I will mention PR to fix the dimension mismatch problem.

**Comment by funnyday16** [comment]
> In any case, the problem of `(pred * weight).sum()` error still needs to be fixed, I will mention PR to fix the dimension mismatch problem.

When I modify it in this way, there will be no dimension mismatch problems, but loss becomes nan as @dave-andersen

**Comment by shinya7y** [comment]
Please use the latest code after #4978.
If still nan, the cause will be out of scope of this issue (e.g., instability of single-stage detectors).
